From f6412cc35e5ed2158c6726e99ae2a01df595e9b8 Mon Sep 17 00:00:00 2001
From: John Hawthorn <john@hawthorn.email>
Date: Sun, 2 Jun 2024 23:40:24 -0700
Subject: [PATCH] Avoid unnecessary writes to imemo_env during GC

Similar to the previous commit, to avoid unnecessary Copy-on-Write
memory use we should only set this flag when it has not previously been
set.

(cherry picked from commit 520ab22725124ef3d588c9ec664b1adfc98da905)

Note: upstream the imemo_env marking lives in imemo.c; in Ruby 3.3 it is
in gc.c (gc_mark_imemo), so this patch applies the same change there.

Second commit of https://github.com/ruby/ruby/pull/10899, backported to
Ruby 3.3 for https://gitlab.com/gitlab-org/gitlab/-/work_items/634879.
---
 gc.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/gc.c b/gc.c
index 0346812f42..a334996996 100644
--- a/gc.c
+++ b/gc.c
@@ -7228,7 +7228,9 @@ gc_mark_imemo(rb_objspace_t *objspace, VALUE obj)
                 GC_ASSERT(env->ep[VM_ENV_DATA_INDEX_ENV] == obj);
                 GC_ASSERT(VM_ENV_ESCAPED_P(env->ep));
                 rb_gc_mark_values((long)env->env_size, env->env);
-                VM_ENV_FLAGS_SET(env->ep, VM_ENV_FLAG_WB_REQUIRED);
+                if (!VM_ENV_FLAGS(env->ep, VM_ENV_FLAG_WB_REQUIRED)) {
+                    VM_ENV_FLAGS_SET(env->ep, VM_ENV_FLAG_WB_REQUIRED);
+                }
                 gc_mark(objspace, (VALUE)rb_vm_env_prev_env(env));
                 gc_mark(objspace, (VALUE)env->iseq);
             }
-- 
2.52.0

